docs/DesignDocs/SGX_QuoteVerify_Integration.md

```diff
@@ -0,0 +1,488 @@
+  and only call `oe_parse_sgx_endorsements()` if caller provides an endorsement buffer
+- In API `_verify_report()` and relevant internal APIs, such as
+  `oe_verify_report_internal()` and `oe_verify_sgx_quote()`, remove the code that
+  loads QPL and retrieves endorsements.
+
+### Add function in enclave-side plugin for QvE Identity verification
+
+For enclave-side verifier plugin, after quote verification, it needs to verify that the result
+comes from a trusted QvE.
+
+Intel® SGX SDK provides a library named `sgx_dcap_tvl` to help the verifier to
+verify QvE's identity, you can refer to [source file](https://github.com/intel/SGXDataCenterAttestationPrimitives/blob/master/QuoteVerification/dcap_tvl/sgx_dcap_tvl.cpp)
+
+This library uses hardcoded QvE identity values, because:
+- There is no need for x509 and JSON parser in verifier enclave's TCB.
+- Most of QvE's identity attributes only change infrequently.
+
+The only identity info may change more often is QvE's ISV SVN, so the verifier needs
+to provide a SVN number as threshold, only when current QvE's ISV SVN is equal to
+or larger than this threshold, the verifier can trust the QvE verification result.
+The verifier can get latest QvE ISV SVN from [QvE Identity at Intel® Provisioning
+Cert Server](https://api.portal.trustedservices.intel.com/documentation#pcs-qve-identity-v2)
+
+As OE SDK cannot use Intel® SGX SDK trusted library directly, we need to port
+this library into OE SDK as part of enclave-side plugin.
+
+```C
+oe_result_t oe_verify_qve_report_and_identity(
+        const uint8_t *p_quote,
+        uint32_t quote_size,
+        const sgx_ql_qe_report_info_t *p_qve_report_info,
+        time_t expiration_check_date,
+        uint32_t collateral_expiration_status,
+        sgx_ql_qv_result_t quote_verification_result,
+        const uint8_t *p_supplemental_data,
+        uint32_t supplemental_data_size,
+        sgx_isv_svn_t qve_isvsvn_threshold)
+```
+All the data returned by the QvE are included in the QvE report data field for
+integrity protection.
+```C
+QvE report_data = SHA256([nonce || quote || expiration_check_date ||
+expiration_status || verification_result || supplemental_data] || 32 - 0x00)
+```
+
+The flow of the API function `oe_verify_qve_report_and_identity()` is as below:
+- Verify QvE report.
+- Verify report data (report data described as above).
+- Check QvE Identity by comparing QvE report fields against hardcode values.
+  - Check that Report.MRSIGNER equals to Hardcoded QvE MRSIGNER.
+  - Check that Misc Select, Attribute and ProdID equal to the hardcoded values,
+    Misc select and Attribute need to apply Mask before comparison.
+- Check that Report.ISVSVN >= Hardcoded ISV SVN.
+
+Options to provide QvE ISV SVN threshold:
+- Option 1: Hardcode QvE ISV SVN in enclave-side plugin. Everytime QvE
+            ISV SVN is updated, the plugin library implementation will need to
+            be updated to stay in sync.
```

Comment:
Do we have a strategy for keeping this logic in sync with the QVL/QVE releases? The ship vehicle is not explicitly mentioned (I expect it's part of the dcap_ql package) so it might be worth pointing out whether the package versioning dependency is separate from the QPL. 

I'll note that this recommendation seems to contradict an earlier desire for the QVL to handle endorsements so that OE doesn't have to keep in sync with the Intel changes to endorsements (which seem less frequent that updates to the libraries?)

Reply:
For the QvE identity verification trusted library, basically we don't need to keep sync with QVL/QvE release. Because nearly all the info in QvE identity would NOT change, except there is a critical issue, such as QvE signing key is leaked.  
But note that QvE ISV SVN is the only exception, it may change from time to time. 

Below QvE identity for example, you will see that except `isvsvn`, all other items are about QvE attribute and mrsigner.
```C
"miscselect":"00000000","miscselectMask":"FFFFFFFF","attributes":"01000000000000000000000000000000","attributesMask":"FBFFFFFFFFFFFFFF0000000000000000","mrsigner":"8C4F5775D796503E96137F77C68A829A0056AC8DED70140B081B094490C57BFF","isvprodid":2,"tcbLevels":[{"tcb":{"isvsvn":3},"tcbDate":"2019-12-11T00:00:00Z","tcbStatus":"UpToDate...
```

In addition, all other SGX endorsements, such as PCK Cert, CRL, TCB info, etc. The internal fields may change from time to time, that's why I propose to get rid of the complex parsing tasks in OE.